src/hotspot/share/opto/loopopts.cpp

```diff
@@ -1392,7 +1392,8 @@ void PhaseIdealLoop::split_if_with_blocks_post(Node *n) {
       Node *prevdom = n;
       Node *dom = idom(prevdom);
       while (dom != cutoff) {
-        if (dom->req() > 1 && dom->in(1) == bol && prevdom->in(0) == dom) {
+        if (dom->req() > 1 && dom->in(1) == bol && prevdom->in(0) == dom &&
+            safe_for_if_replacement(dom)) {
```

Comment:
We skip replacement for counted loop but we still continue searching IDOMs. Which may find another dominating test before loop and replace node's control which may be not correct since we skipped loop with the same test.
And you said "The fix I propose is to detect that corner case and
prevent the If to be replaced in that case." We should bailout then.

Reply:
Hi Vladimir. Thanks for reviewing this.
I'm not sure I understand. If dom is a CountedLoopEnd and dom->in(1) == bol then that test depends on the iv Phi so cutoff is the CountedLoop node and that code is not going to move above the loop.

Reply:
Good point. And we can't have the same test (same Bool) for an other test inside loop  - it would be an other loop.